includes/core-changes/codeanalysis/5.0/ca1416-platform-compatibility-analyzer.md

```diff
@@ -0,0 +1,127 @@
+
+#### Examples
+
+- The <xref:System.Console.Beep(System.Int32,System.Int32)?displayProperty=nameWithType> method is only supported on Windows (it's decorated with `[SupportedOSPlatform("windows")]`). The following code will produce a CA1416 warning at build time if the project [targets](../../../../docs/standard/frameworks.md) `net5.0` (but not `net5.0-windows`). For actions you can take to avoid the warning, see [Recommended action](#recommended-action).
+
+  ```csharp
+  public void PlayCMajor()
+  {
+      Console.Beep(261, 1000);
+  }
+  ```
+
+- The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Blazor.WebAssembly.Sdk">`) or includes `browser` as a supported platform (`<SupportedPlatform Include="browser" />`) in the project file.
+
+  ```csharp
+  public void CreateImage()
+  {
+      Image newImage = Image.FromFile("SampImag.jpg");
+  }
+  ```
+
+#### Version introduced
+
+5.0 RC1
+
+#### Recommended action
+
+Ensure that platform-specific APIs are only called when the code is running on an appropriate platform. You can achieve this at build time using preprocessor directives, or at run time by using the platform guards in the <xref:System.OperatingSystem?displayProperty=nameWithType> class:
+
+- Add a `#if` [preprocessor directive](../../../../docs/csharp/language-reference/preprocessor-directives/preprocessor-if.md) around platform-specific API calls:
+
+  ```csharp
+  #if NET50_WINDOWS
+      Console.Beep(261, 1000);
+  #endif
+  ```
+
+- Check the current operating system using one of the <xref:System.OperatingSystem?displayProperty=nameWithType> platform-checking methods, for example, `System.OperatingSystem.IsWindows()`, before calling a platform-specific API. You can use one of these methods in the condition of an `if` statement:
+
+  ```csharp
+  public void PlayCMajor()
+  {
+      if (OperatingSystem.IsWindows())
+      {
+          Console.Beep(261, 1000);
+      }
+  }
+  ```
+
+  ...or pass its result to <xref:System.Diagnostics.Debug.Assert(System.Boolean)?displayProperty=nameWithType>:
+
+  ```csharp
+  public void PlayCMajor()
+  {
+      Debug.Assert(OperatingSystem.IsWindows());
+      Console.Beep(261, 1000);
+  }
+  ```
```

Comment:
Shouldn't the compiler omits the call to Debug.Assert in a release build and produces the warning for this? i.e. this snippet should fix the warning only for a debug build.

Do you know who can confirm?

I also cannot see IsWindows method in the current API reference. Is it missing or not shipped yet?

Reply:
IsWindows() is an RC1 method. We haven't updated the API docs yet for RC1.

For Debug.Assert(), I think the analyzer is smart enough to detect it, even for a release build. @buyaa-n can confirm.

Reply:
> Shouldn't the compiler omits the call to Debug.Assert in a release build and produces the warning for this? i.e. this snippet should fix the warning only for a debug build.

Yes, it is trimmed for release build and that is the whole point, we don't want it fail runtime.  If dev doesn't want to add new if check and make a call to OperatingSystem.IsWindows() at runtime they could use it

Reply:
@gewarren you might want to add this `Using Debug.Assert allows the check to be trimmed out of release builds`

Reply:
Oh it was added below, never mind 😄